# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
# Script to assemble methylation  table for Daniel -----
# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@


# ++++++++++++++++++++
# Dataframes and Lists
# ++++++++++++++++++++

# biospecimen_cl$sample:
#    dataframe with all metadata taken from GDC
# biospecimen_cl:
#    list of dataframes regariding biospecimens
# clinical:
#    list of tabs regarding clinical data
# icgc:
#    dataframe with metadata of the 40 sample of wgs from ICGC
# wgs_specimens :
#   specimen ids of the 40 wgs samples
# metadata:
#    dataframe of metadata from Biolink/ GCD legacy
# tab_dan /tab_dan_no_ben:
#   reshaped and cleaned dataframe for Daniel Biolink/ GCD legacy with or without benign (ben)
#   samples. the replicates were removed
# tab_dan_wgs
#   methilation table for Daniel with WGS information updated
# tab_dan_wgs_rna
#   methilation table for Daniel with RNA information updated
# wgs_table_us
#   WGS table for Daniel


# @@@@@@@@@@@@@@@@@@@@@@@@
# Load TCGA Metadata -----
# @@@@@@@@@@@@@@@@@@@@@@@@
# load obj generated by here("code", "get_metadata_tcga.R")
library(here)
source(here('code', "libraries.R"))

load(file = here("objs", "annotations_methylation.RDA"))

sum(apply(biospecimen_cl$sample, 1, function(x) all(is.na(x))))



# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
# Methylation  -----
# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

# This is to get metadata of the methylation samples
# getGDCprojects()$project_id
#
# query <- GDCquery(project = "TCGA-PRAD",
#          data.category = "Raw microarray data",
#          data.type = "Raw intensities",
#          experimental.strategy = "Methylation array",
#          legacy = TRUE,
#          file.type = ".idat",
#          platform = "Illumina Human Methylation 450")
#
#
# metadata <- getResults(query)
# save(metadata, file = here("data", "methylation", "metadata.RDA"))

load(file = here("data", "methylation", "metadata.RDA"))
metadata


# @@@@@@@@@@@@@@@@@
# WGS -----
# @@@@@@@@@@@@@@@@@

# This table will need to have info also on matching methylation and RNA
# Identify WGS donors and WGS specimens
# the identifier were taken from and a list is also in here("data", "dna_whole", "ICGC_donors.csv")
# https://dcc.icgc.org/repositories?filters=%7B%22file%22:%7B%22projectCode%22:%7B%22is%22:%5B%22PRAD-US%22%5D%7D,%22primarySite%22:%7B%22is%22:%5B%22Prostate%22%5D%7D,%22study%22:%7B%22is%22:%5B%22PCAWG%22%5D%7D,%22experimentalStrategy%22:%7B%22is%22:%5B%22WGS%22%5D%7D,%22fileFormat%22:%7B%22is%22:%5B%22BAM%22%5D%7D%7D%7D&files=%7B%22from%22:1,%22size%22:25%7D

wgs_donors <- list(
  "TCGA-CH-5763",
  "TCGA-EJ-5506",
  "TCGA-HI-7169",
  "TCGA-CH-5788",
  "TCGA-EJ-5503",

  "TCGA-HC-8258",
  "TCGA-HC-7233",
  "TCGA-G9-6336",
  "TCGA-HC-7079",
  "TCGA-EJ-7791",

  "TCGA-HC-7737",
  "TCGA-CH-5771",
  "TCGA-G9-7522",
  "TCGA-CH-5789",
  "TCGA-HC-7744",

  "TCGA-G9-6370",
  "TCGA-G9-6365",
  "TCGA-HC-7740",
  "TCGA-HC-7075",
  "TCGA-CH-5750"
)

wgs_controls <- list(
  "TCGA-CH-5763-11A",
  "TCGA-EJ-5506-10A",
  "TCGA-HI-7169-10A",
  "TCGA-CH-5788-10A",
  "TCGA-EJ-5503-10A",

  "TCGA-HC-8258-10A",
  "TCGA-HC-7233-10A",
  "TCGA-G9-6336-10A",
  "TCGA-HC-7079-10A",
  "TCGA-EJ-7791-10A",

  "TCGA-HC-7737-11A",
  "TCGA-CH-5771-11A",
  "TCGA-G9-7522-10A",
  "TCGA-CH-5789-10A",
  "TCGA-HC-7744-10A",

  "TCGA-G9-6370-10A",
  "TCGA-G9-6365-10A",
  "TCGA-HC-7740-10A",
  "TCGA-HC-7075-10A",
  "TCGA-CH-5750-10A"
)

# we have 17 wgs samples that are from a meth donor but belong to different samples (blood).
# we have to add those and indicate that they are control

wgs_tumors <- paste0(wgs_donors, "-01A")
wgs_specimens <- unlist(c(wgs_controls, wgs_tumors))

# checking the type of wgs sample
biospecimen_cl$sample %>%
  filter(sample_submitter_id %in% wgs_specimens) %>%
  select(sample_submitter_id, sample_type) %>%
  distinct(sample_submitter_id, .keep_all = TRUE)


# these are the samples that do not match between methylation and wgs
# so these are the one with blood
samples_not_meth <- wgs_specimens[!wgs_specimens %in% metadata$sample.submitter_id]

# These are the replicate that we want to remove because of low tumor content
replicates_remove <- c(
  "TCGA-HC-7740-01B",
  "TCGA-HC-8258-01B",
  "TCGA-HC-8261-01A",
  "TCGA-HC-8265-01A"
  )

benigns_remove <-  c(
  "TCGA-CH-5769-11",
  "TCGA-G9-6362-11",
  "TCGA-G9-6363-11",
  "TCGA-G9-6499-11"
)



save(wgs_donors, wgs_specimens, replicates_remove, file = here("objs", "wgs_donors_specimens.RDA"))


tab_dan <-
  metadata %>%
 # add column for type of file
  mutate(idat = if_else(str_detect(file_name, "Red"), "red", "green")) %>%
  select(cases.submitter_id,
         sample.submitter_id,
         idat, file_name,
         sample_type,
         platform, updated_datetime) %>%

  # column type of file in long format
  pivot_wider(names_from = idat, values_from = c("file_name", "updated_datetime")) %>%

  # remove bad samples
    filter(!sample.submitter_id %in% replicates_remove) %>%

  # these are the BEN that are contaminated according to Clarissa
    filter(!sub(".$", "", sample.submitter_id) %in% benigns_remove) %>%


    mutate(
         Country = "USA",
         # unique donor id
         PPCG_Donor_ID = paste0("NEW_", as.numeric(as.factor(cases.submitter_id))),

         # check if we have matching wgs sample
         Matching_WGS_Sample = if_else(sample.submitter_id %in% wgs_specimens,
                                       "ADD",
                                       NA_character_ ),
         WGS_Local_Sample_ID = if_else(sample.submitter_id %in% wgs_specimens,
                                       sample.submitter_id,
                                       NA_character_ ),
         PPCG_Meth_Sample_ID = NA_character_,
         Platform = "450K",
         Meth_Local_Sample_ID = sample.submitter_id,
         Meth_Sample_Type = case_when(
           sample_type == "Solid Tissue Normal" ~ "BEN",
           sample_type == "Primary Tumor" ~ "TUM",
           sample_type == "Metastatic" ~ "MET"
         ),
         MetastaticSite = NA_character_, # change

         # if the sample is among the one used for wsg then it is a different portion of the same sample
         Meth_Distance_To_WGS_Sample = if_else(
           sample.submitter_id %in% wgs_specimens,
           "MATCHED",
           NA_character_ ),

         Meth_RNAseq_annot = NA_character_,
         idat_grn_filename = file_name_green,
         idat_red_filename = file_name_red,
         Date_of_scan_or_Sequencing = NA_character_
         ) %>%
  select(cases.submitter_id, sample.submitter_id, Country:Date_of_scan_or_Sequencing)


# @@@@@@@@
# RNA info
# @@@@@@@@

# Identify RNA donors and RNA specimens
rna_donors <- biospecimen_cl$analyte  %>%
  filter(analyte_type == "RNA") %>%
  distinct(case_submitter_id) %>%
  unlist()

rna_specimens <- biospecimen_cl$analyte  %>%
  filter( analyte_type == "RNA") %>%
  distinct(sample_submitter_id) %>%
  unlist()

tab_dan_wgs_rna <- tab_dan %>%
  # if the same donor then we have "at least RNA seq that is distant
  mutate(
    Meth_RNAseq_annot = if_else(cases.submitter_id %in% rna_donors, "DISTANT", NA_character_)

  ) %>%
  # if it is from same sample then it is matched
  mutate(
    Meth_RNAseq_annot = if_else(
      sub(".$", "", sample.submitter_id) %in% sub(".$", "", rna_specimens),
      "MATCHED",
      Meth_RNAseq_annot
      ))

View(tab_dan_wgs_rna)

# @@@@@@@@@@@@@@@@@@@@@@@@@@
# Some info on RNA  metadata
# @@@@@@@@@@@@@@@@@@@@@@@@@@

# Unique Donors 498
unique(metadata$cases.submitter_id)

# Unique samples 553
unique(metadata$sample.submitter_id)

tab_dan_cl%>%
  group_by(sample.submitter_id) %>%
  View()

# Tecnical replicates
sample_donors <- metadata %>%
  group_by(cases.submitter_id, sample_type) %>%
  summarize(n = n()/2) %>%
  arrange(desc(n))

# these are tecnical replicates
sample_donors %>%
  filter(n == 2) %>%
  distinct(cases.submitter_id) %>%
  dput()

# these are different type of samples from same patients
sample_donors %>%
  filter(n != 2) %>%
  group_by(cases.submitter_id) %>%
  summarise(n = n()) %>%
  arrange(desc(n))  %>%
  filter(n > 1) %>%
  View()


# donors with WGS
metadata %>%
  filter(cases.submitter_id %in% !!wgs_donors) %>%
  group_by(cases.submitter_id) %>%
  summarize(n = n()/2) %>%
  arrange(desc(n))


# @@@@@@@@@@@@@@@@@@
# Some info last tab
# @@@@@@@@@@@@@@@@@@

# Unique Donors 498
unique(tab_dan_wgs_rna$cases.submitter_id)

nrow(tab_dan_wgs_rna)

# Unique samples 545
unique(tab_dan_wgs_rna$sample.submitter_id)


# Multiple samples from same donor
mult_sample <- tab_dan_wgs_rna %>%
  group_by(cases.submitter_id, Meth_Sample_Type) %>%
  summarize(n = n()) %>%
  pivot_wider(names_from = "Meth_Sample_Type", values_from = "n", values_fill = 0) %>%
  mutate(tot = sum(TUM, BEN, MET)) %>%
  arrange(desc(tot))

mult_sample %>%
  filter(tot > 1) %>%
  nrow()

mult_sample %>%
  filter(tot > 1) %>%
  View()

# sample with WGS: 23   (17 blood are not included yet)
tab_dan_wgs_rna %>%
  filter(sample.submitter_id %in% wgs_specimens) %>%
  nrow()


donor_extra_sample <- tab_dan_wgs_rna %>%
  group_by(cases.submitter_id) %>%
  summarize(n = n()) %>%
  arrange(desc(n)) %>%
  filter(n > 1) %>%
  filter(!cases.submitter_id %in% sub("-.{3}$", "", samples_not_meth)) %>%
  pull(cases.submitter_id )

tab_dan_wgs_rna_cl <- tab_dan_wgs_rna %>%
  select(! 1:2)

write.csv(tab_dan_wgs_rna_cl, file = here("reports", "meth_tab_us.csv"), row.names = FALSE )
saveRDS(tab_dan_wgs_rna_cl, file = here("objs", "meth_tab_us.RDS"))


# 498 unique donors with TUM samples +
# 46 benign samples
# 1 metastatic from "TCGA-V1-A9O5"
498 + 46 + 1 == 545

# sample with matching WGS: 23   (17 blood are not included yet)
# 17 blood not included


wgs_files <- read.csv(here("data", "dna_whole", "ICGC_donors.csv"))

col_names_wgs <- c(
  "Country",
  "Donor_UUID",
  "PPCG_Donor_ID",
  "PPCG_Sample_ID",
  "Local_Sample_ID",
  "Local_Donor_ID",
  "Sanger_UUID",
  "EO_UUID",
  "Sample_UUID",
  "Matched_Normal_PPCG_Sample_ID",
  "Is_Normal",
  "Is_Normal_for_Donor",
  "Sample_Type",
  "Progether_ID"
)

wgs_table <- data.frame(matrix(ncol = 14, nrow = length(wgs_specimens)))
names(wgs_table) <- col_names_wgs

wgs_table_us <-
  wgs_table %>%
  mutate(Country = "USA", Local_Sample_ID = !!wgs_specimens) %>%

  mutate(Local_Donor_ID = sub("-.{3}$", "", Local_Sample_ID)) %>%

  inner_join(biospecimen_cl$sample[, c("sample_submitter_id", "sample_type")],
             by = c("Local_Sample_ID" = "sample_submitter_id")) %>%
  mutate(Sample_Type = sample_type) %>%
  select(-sample_type) %>%
  mutate(Sample_Type = case_when(
    Sample_Type == "Solid Tissue Normal" ~ "Solid Tissue Normal (Benign)",
    Sample_Type == "Primary Tumor" ~ Sample_Type,
    Sample_Type == "Blood Derived Normal" ~ "Blood Derived Normal (Germline)",

  )) %>%
  inner_join(wgs_files[, c("tcga_sample_id", "ICGC.Donor", "Sample.ID", "Object.ID")],
            by = c("Local_Sample_ID" = "tcga_sample_id")) %>%
  mutate(Sample_UUID = Object.ID) %>%
  select(-Object.ID)



write.csv(wgs_table_us, file = here("reports", "wgs_table_us.csv"), row.names = FALSE )


# @@@@@@@@@@@@@@@@@@@@@@@@@@@
# Add Sample ID ICGC in table
# @@@@@@@@@@@@@@@@@@@@@@@@@@@

tab_dan_wgs_rna_cl  <- tab_dan_wgs_rna %>%
  select(!1:2) %>%
  left_join(wgs_files[, c("tcga_sample_id", "Sample.ID")],
             by = c("WGS_Local_Sample_ID" = "tcga_sample_id"))

View(tab_dan_wgs_rna_cl)

tab_dan_wgs_rna_cl %>%
  write.csv(file = here("reports", "meth_tab_us.csv"), row.names = FALSE )

# WGS ids
# we are adding back the ppcg ids that daniel gave us
ppcg_wgs_ids <- read.csv(file = here("data", "ppcg_wgs_ids.csv"))

ppcg_wgs_ids_for_join <- ppcg_wgs_ids %>%
  select(PPCG_Donor_ID:Local_Donor_ID)

names(ppcg_wgs_ids_for_join) <- paste0("wgs_dan_", names(ppcg_wgs_ids_for_join))


tab_dan_wgs_rna_PPCG <- tab_dan_wgs_rna_cl %>%
  left_join(ppcg_wgs_ids_for_join, by = c("WGS_Local_Sample_ID" = "wgs_dan_Local_Sample_ID")) %>%

  # if we do not have a WGS id we keep the new
  mutate(PPCG_Donor_ID = if_else(is.na(wgs_dan_PPCG_Donor_ID), PPCG_Donor_ID, wgs_dan_PPCG_Donor_ID)) %>%
  mutate(Matching_WGS_Sample = wgs_dan_PPCG_Sample_ID)


tab_dan_wgs_rna_PPCG  %>%
  write.csv(file = here("reports", "meth_tab_us.csv"), row.names = FALSE )

save(tab_dan_wgs_rna_PPCG, ppcg_wgs_ids, file = here("objs", "methylation_wgs_tabs.RDA"))



#   # 498 unique donors with TUM samples +
#   # 46 benign samples
#   # 1 metastatic from "TCGA-V1-A9O5"
#   498 + 46 + 1 = 545
#
# # sample with matching WGS: 23   (17 blood are not included yet)
# # 17 blood not included


# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
# Methylation: some clinical data----
# @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

# This is to create the Clinical data needed by Clarissa
# We load the meth ids as provided by Daniel
meth_usa <- read_excel("data/Meth_USA_PPCGIDs.xlsx")

# legacy
clinical <-
  read.delim(here("data", "prad_tcga_legacy", "data_bcr_clinical_data_patient.txt"),
             header = TRUE,
             comment.char = "#")


names(clinical) <- tolower(names(clinical))

# https://wiki.oicr.on.ca/display/PPCP/Clinical+data+dictionary?preview=%2F146836685%2F160271629%2FPanprostate_Clinical_Data_Dictionary_18_05_2021.docx
clinical_info <- clinical %>%
  # select()
  rename(local_donor_id = patient_id,
         # `ppcg_donor_id`
         age_at_tumour_collection = age,
         # PSA_at_tumour_collection = psa_most_recent_results,
         clin_t_stage = clin_t_stage,
         path_t_stage = path_t_stage ,
         # `de novo mets`  not sure
         path_gleason_sum = gleason_score,
         gleason_predominantly = gleason_pattern_primary,
         gleason_less_predominantly = gleason_pattern_secondary,
         death_ind = os_status,
         donor_interval_of_last_followup = os_months,
         cancer_specific_death_ind = patient_death_reason, # (cancer death; non-cancer death; alive)
         relapse_ind = biochemical_recurrence_indicator,
         donor_relapse_interval = days_to_biochemical_recurrence_first,
         # mets_ind (mets or no mets)
         # donor_first_mets_interval_from_M0
         `Bone_metastasis` = bone_scan_result,
         time_to_bone_mets = days_to_bone_scan_performed,
         # LN_metastasis
         # time_to_LN_mets
         # Organ_metastasis
         # time_to_organ_mets
         # location_first_mets_comment
         # metastatic_sites_comment = tissue_source_site
         ) %>%
  select(local_donor_id,
         age_at_tumour_collection,
         clin_t_stage,
         path_t_stage,
         path_gleason_sum,
         gleason_predominantly,
         gleason_less_predominantly,
         death_ind,
         donor_interval_of_last_followup,
         cancer_specific_death_ind, # (cancer death; non-cancer death; alive)
         relapse_ind,
         donor_relapse_interval,
         # mets_ind (mets or no mets)
         # donor_first_mets_interval_from_M0
         `Bone_metastasis`,
         time_to_bone_mets) %>%  # this is in day
  mutate(death_ind = recode(death_ind,
                            "0:LIVING" = "alive",
                            "1:DECEASED" = "dead"
  )) %>%

  mutate(
    cancer_specific_death_ind = recode(cancer_specific_death_ind,
                            "[Not Available]" = NA_character_,
                            "Prostate Cancer" = "cancer death",
                            "non-cancer death" = "non-malignant disease",
                            "non-cancer death" = "Other",
  )) %>%
  mutate(cancer_specific_death_ind =
           if_else(death_ind == "alive", "alive",cancer_specific_death_ind)) %>%

  # ASK LUIGI < ---------

  mutate(
    Bone_metastasis = recode(Bone_metastasis,
                                       "Normal (no evidence of prostate cancer) [cM0]" = "no",
                                       "Equivocal" = NA_character_,
                                       "[Not Available]" =  NA_character_,
                                       "Abnormal (not related to prostate cancer)" = "no",
                                       "Prostate Cancer Metastases Present [cM1b]" = "yes"
    ))


metadata_methilation <- meth_usa %>%
  select(PPCG_Donor_Id, PPCG_Meth_Sample_ID, Meth_Local_Sample_ID) %>%
  mutate(local_donor_id = gsub(".{4}$", "", Meth_Local_Sample_ID)) %>%
  left_join(clinical_info)


write.csv(metadata_methilation, here("reports", "metadata_methylation.csv"))





















